chapters/os_tooling.md

```diff
@@ -0,0 +1,84 @@
+# Offline Compilation of OpenCL Kernel Sources
+
+Aside from online compilation during application execution, OpenCL kernel sources can be compiled offline into binaries that can be loaded into the drivers using special API calls (e.g. `clCreateProgramWithBinary` or `clCreateProgramWithIL`).
+
+This section describes available open source tools for offline compilation of OpenCL kernels.
+
+## Open Source Tools
+
+* [clang](https://clang.llvm.org/) is a compiler front-end for the C/C++ family of languages including OpenCL C and C++ for OpenCL that can produce executable binaries (e.g. AMDGPU), or portable binaries (e.g. SPIR). It is part of [the LLVM compiler infrastructure project](https://llvm.org/), and there is information regarding [OpenCL kernel language support and standard headers](https://clang.llvm.org/docs/UsersManual.html#opencl-features).
+* [libclc](https://libclc.llvm.org/) is a generic and portable implementation of OpenCL builtin function libraries for OpenCL 1.1 - and some functions from later versions of OpenCL can be found there too.
+* [SPIRV-LLVM Translator](https://github.com/KhronosGroup/SPIRV-LLVM-Translator) provides a library and the __llvm-spirv__ tool for bidirectional translation between LLVM IR and SPIR-V. 
+* [clspv](https://github.com/google/clspv) compiler and [clvk](https://github.com/kpet/clvk) runtime layer enable OpenCL applications to be executed with Vulkan drivers.
+* [SPIR-V Tools](https://github.com/KhronosGroup/SPIRV-Tools) provide a set of utilities to process SPIR-V binaries including __spirv-opt__ optimizer, __spirv-link__ linker, __spirv-dis__/__spirv-as__ (dis-)assembler, and __spirv-val__ validator.
+
+### Compiling Kernels to SPIR-V
+
+The open source tools can be used to perform full compilation from OpenCL kernel sources into SPIR-V.
+
+<p align="center">
+<br>
+<img src="../images/opencl_to_spirv_tooling.jpg">
+<br> <br>
+  <b>Offline compilation flow for OpenCL kernels into SPIR-V</b>
+<br> <br>
+</p>
+
+#### Examples
+
+If you want to try the above compilation flow for yourself, after installing the tools, you can use the following commands.
+
+##### Compile for OpenCL runtime
+
+__(i)__ Compiling OpenCL C/C++ for OpenCL file into intermediate LLVM IR (for 32 bit targets) formats:
+
+```
+clang -c -target spir -O0 -emit-llvm -o test.bc test.cl
```

Comment:
Does it make sense to have some Compiler Explorer links to demonstrate these command lines?

Reply:
Yes this is a good idea. The only issue is that the exact command-lines won't be the same since Compiler Explorer doesn't support OpenCL natively. :( 

So we could need to add -x flag everywhere. Do you think it is ok?

Reply:
I will resolve this for now. We can always add those in a separate PR. :)